content/post/scaling-workloads/index.md

```diff
@@ -0,0 +1,116 @@
+This results in a program which looks like this.
+
+```asm
+add ax, 1
+```
+
+This program results in a single clock cycle and takes the least amount of execution time compared
+to the other variants of this program. However, the optimization of removing the `mov` to `ax` can
+be dangerous. If `ax` is not `1` then it would result in a different result being stored into `ax`.
+This is an example of an optimization which can be dangerous. It is considered dangerous as it
+is assuming the CPU state being a certain way with no checks. Adding checks can result in degrated
+performance. This is why many program which deals with high performance may use hand written assembly.
+The compiler may introduce certain optimizations such as the one in this example that either produces
+undefined behavior or is not optimal enough. Luckily, most people do not have to deal with having to
+understand what is going on. Modern compilers like GCC and LLVM's clang are optimal enough to prevent
+most issues. However, the thing which is important is the size of the workloads, having a good CPU for
+the workload, and enough RAM for the workload. Having the right CPU also matters as well, it is better
+to have a CPU for a specific architecture instead of cross compiling. This is because you have less
+to bootstrap for a specific CPU architecture. You typically already have what you need to compile for
+your native architecture.
+
+Having enough RAM depends on what the system typically does and how much you can parallelize really are
+the most important things here. This is going to become more apparant with the world moving away from x86
+and to ARM. ARM is quite different, it is known as a RISC (Reduced Instruction Set Computing) architecture.
+x86 is considered a CISC (Complex Instruction Set Computing) architecture. x86 has many more instructions
+compared to ARM. However, many instructions in a CISC architecture are similar to each other. This means
+you may have 30 different move instructions which each are optimized for a different data type. While on ARM,
+you have just what is needed. There can be just 5 move instructions on a RISC compared to the 30 in a CISC.
+This results in less complexity in the CPU which means less heat generated due to the less amount of silicon
+being used. This can result in a CPU operating colder and less prone to thermal throttling, it also requires
+less cooling. This can be optimal for laptops and phone, in fact this is why ARM is used in phones. In my
+experience being a modern ARM user (Apple M1 Pro), the ARM processor is actually faster than x86 (Ryzen 5
+3600 on my desktop). However, despite having less optimized instructions, it is actually more beneficial
+to have more CPU cores. This is because you can only execute instructions so fast despite what the clock
+rate is set at. This has to do with how the silicon is designed and its thermal limits. Naturally, x86
```

Comment:
This conclusion does tie together the explanations in previous paragraphs but I don't feel properly develops a conclusion about parallelization from those explanations.